### PR TITLE
[FrameworkBundle] Document the AbstractController

### PR DESCRIPTION
**Pull request by GuilhemN**
Document https://github.com/symfony/symfony/pull/18193. 
I'm not sure it should be merged right now though as it's an experimental feature.

\cc @dunglas

### CODE REVIEW

**Comment by HeahDude** [comment]
@GuilhemN Would you like to document the abstract controller introduced in https://github.com/symfony/symfony/pull/22157 instead? Thanks!

**Comment by GuilhemN** [comment]
@HeahDude oh that's already merged, I didn't expect it so quickly. I'll make the changes :)

**Comment by GuilhemN** [comment]
Updated to document the `AbstractController` instead.

ping @nicolas-grekas

**Comment by HeahDude** [comment]
Closes #7705.

File: controller/service.rst [hunk 1]
```diff
@@ -135,14 +201,15 @@ the route ``_controller`` value:
...
+Alternatives to Controller shortcut Methods
+-------------------------------------------
 
-When using a controller defined as a service, it will most likely not extend
-the base ``Controller`` class. Instead of relying on its shortcut methods,
-you'll interact directly with the services that you need. Fortunately, this is
-usually pretty easy and the base `Controller class source code`_ is a great
-source on how to perform many common tasks.
+In case you don't want to inject the container in your controllers,
+it's possible to get ride of controller shortcut methods: you can interact
```
Comment (nicolas-grekas) [inline controller/service.rst]: to get rid of  (no "e")

File: controller/service.rst [hunk 2]
```diff
@@ -83,6 +81,74 @@ Then you can define it as a service as follows:
...
+With the following configuration:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # app/config/services.yml
+        services:
+            _instanceof:
+                Symfony\Bundle\FrameworkBundle\Controller\AbstractController:
```
Comment (nicolas-grekas) [inline controller/service.rst]: this block is not needed, let's remove it from the doc

Comment (nicolas-grekas) [inline controller/service.rst]: In fact, there is no config to do at all, an AbstractController doesn't need to be a service at all

File: controller/service.rst [hunk 3]
```diff
@@ -83,6 +81,74 @@ Then you can define it as a service as follows:
 
         $container->register('app.hello_controller', HelloController::class);
 
+Using the controller shortcuts
+------------------------------
+
+To use the traditional controller shortcuts, you can use the
```
Comment (nicolas-grekas) [inline controller/service.rst]: this paragraph is a bit strange to me, it does really teach enough to the reader imho.
eg
> if you prefer explicit injection (you can't access the container)

"why would I prefer explicit injection? esp when the words in the brackets suggest that this is less powerful, so let's just to the previous `Controller`." Which then raises "why did they tell me about this? do I miss anything?" leading to confusion.

my 2cts :)

File: controller/service.rst (hunk 2 above)
Comment (GuilhemN) [inline controller/service.rst]: Do you think it should be documented in a different section then?

**Comment by GuilhemN** [comment]
Updated to document the `AbstractController` in the `controller.rst` file to not limit its use to controllers as services.

File: controller.rst [hunk 4]
```diff
@@ -116,19 +116,26 @@ For more information on routing, see :doc:`/routing`.
 .. index::
    single: Controller; Base controller class
 
-The Base Controller Class & Services
-------------------------------------
+The Base Controller Classes & Services
```
Comment (xabbuh) [inline controller.rst]: we should add a label for the old headline to not break any deep links

Comment (GuilhemN) [inline controller.rst]: Fixed, I'm not 100% sure this is how it works though.

File: controller.rst [hunk 5]
```diff
@@ -241,10 +251,9 @@ are used for rendering templates, sending emails, querying the database and
 any other "work" you can think of. When you install a new bundle, it probably
 brings in even *more* services.
 
-When extending the base controller class, you can access any Symfony service
+When extending the base ``Controller`` class, you can access any Symfony service
```
Comment (HeahDude) [inline controller.rst]: I think we're missing pros of using the new AbstractController, dependencies are explicit either in a constructor or in action signatures, but the controller needs to be configured to get injection.

Comment (GuilhemN) [inline controller.rst]: Do you mean we have to tell that the controller has to be configured as a service ? I updated the text to reflect that.

Comment (HeahDude) [inline controller.rst]: Hmm it seems I got confused, reading @nicolas-grekas's comment above:
>In fact, there is no config to do at all, an AbstractController doesn't need to be a service at all
The injection in this case is relying on an argument value resolver, so the controller does not need any configuration indeed.

Comment (HeahDude) [inline controller.rst]: What I just said is wrong, the getters has nothing to do with action injections, so I don't get it, this must be configured somehow, it's just already done by the bundle.

**Comment by weaverryan** [comment]
Thanks @GuilhemN! I'll merge this into my #7807 :)

File: controller.rst [hunk 6]
```diff
@@ -116,19 +116,30 @@ For more information on routing, see :doc:`/routing`.
 .. index::
    single: Controller; Base controller class
 
-The Base Controller Class & Services
-------------------------------------
+.. _anchor-name:
+   :ref:`The Base Controller Classes & Services <the-base-controller-class-services>`
```
Comment (weaverryan) [inline controller.rst]: I'm not familiar with this `anchor-name` thing - were you trying to make sure the old anchor still worked? If so, #7867 for an example. I'll update this in my PR

Comment (GuilhemN) [inline controller.rst]: Yes that was [a request from @xabbuh](https://github.com/symfony/symfony-docs/pull/7657#discussion_r109855847).
Thanks for the example, I never did that before and obviously didn't do it well by randomly trying :)

**Pull request merged**